NCsvPerf/CsvReadable/Implementations/RecordParser.cs

```diff
@@ -0,0 +1,145 @@
+﻿using System;
+using System.Buffers;
+using System.Collections.Generic;
+using System.IO;
+using System.IO.Pipelines;
+using System.Linq.Expressions;
+using System.Text;
+
+namespace Knapcode.NCsvPerf.CsvReadable
+{
+    /// <summary>
+    /// Package: https://www.nuget.org/packages/recordparser
+    /// Source: https://github.com/leandromoh/RecordParser
+    /// </summary>
+    public class RecordParser : ICsvReader
+    {
+        private readonly ActivationMethod _activationMethod;
+
+        public RecordParser(ActivationMethod activationMethod)
+        {
+            _activationMethod = activationMethod;
+        }
+
+        public List<T> GetRecords<T>(MemoryStream stream) where T : ICsvReadable, new()
+        {
+            var activate = ActivatorFactory.Create<T>(_activationMethod);
+            var reader = BuildReader(activate);
+            var result = ProcessStream<T>(stream, spanLine =>
+            {
+                var fields = reader.Parse(spanLine);
+                var record = activate();
+
+                record.Read(i => fields[i]);
+
+                return record;
+            });
+
+            return result;
+        }
+
+        private static Expression<Func<string[], string>> buildExpression(int i)
+        {
+            var arrayExpr = Expression.Parameter(typeof(string[]));
+            var indexExpr = Expression.Constant(i);
+            var arrayAccessExpr = Expression.ArrayAccess(arrayExpr, indexExpr);
+
+            return Expression.Lambda<Func<string[], string>>(arrayAccessExpr, arrayExpr);
+        }
+
+        private static global::RecordParser.Parsers.IVariableLengthReader<string[]> BuildReader<T>(Activate<T> activate) where T : ICsvReadable
+        {
+            var columnCount = activate().GetColumnCount();
+            var buffer = new string[columnCount];
+            var builder = new global::RecordParser.Builders.Reader.VariableLengthReaderBuilder<string[]>();
+
+            for (var i = 0; i < buffer.Length; i++)
+                builder.Map(buildExpression(i), i);
+
+            var cache = new Dictionary<int, string>()
+            {
+                [string.GetHashCode(Span<char>.Empty)] = string.Empty
+            };
+
+            builder.DefaultTypeConvert(GetOrAddString);
+
+            var reader = builder.Build(",", factory: () => buffer);
+
+            return reader;
+
+            string GetOrAddString(ReadOnlySpan<char> span)
+            {
+                var key = string.GetHashCode(span);
+
+                return cache.TryGetValue(key, out var text)
```

Comment:
Using just the hashcode of the string as the key into the dictionary will occasionally result in two different strings (which happen to have the same hashcode) to be considered equal. This might never occur with the test data set, since the number of unique strings is relatively small, but this code should be corrected. The default string hashing algorithm in .NET includes per-process randomization, so even with the limited test data set it will likely fail eventually.

You might consider using the `Ben.StringIntern` nuget package for the string caching part, it is very high performance and (I assume) correct. I have an implementation of my own, but I'd probably swap for Ben's now; I would have used it originally, but it didn't exist at the time.

Reply:
@MarkPflug thanks for the point, I didnt know about the possibility of hash colisions, but indeed it was in the method documentation. Fixed in [#b2cfd4f](https://github.com/joelverhagen/NCsvPerf/pull/30/commits/b2cfd4f73b3b9bd3a95d1c27f21e74c659bca0a2)

Reply:
By the way, I saw that your lib Sylvan_Data_Csv and CsvHelper allocates 261 MB while RecordParse and Cursively does 345 MB. I really would like to know how do you allocates less than that, since each parsed object allocates zero on heap in `RecordParse` project. Do you have any guess?

Reply:
> how do you allocates less than that

Interesting question. I don't know the answer. Let's take a look.

The first thing to observe is that there is a baseline allocation that is required by the benchmark. The benchmark requires us to construct a `List<PackageAsset>` containing 1mil instances. I've created a benchmark that measures that minimal allocation:

```
        [Benchmark]
        public void AllocationBaseline()
        {
            var c = 1_000_000;
            var f = new List<PackageAsset>(c);
            for (int i = 0; i < c; i++)
            {
                f.Add(new PackageAsset());
            }
        }
```

Running that benchmark reports an allocation baseline of ~257MB. Is this the absolute minimum that we could allocate? Practically, yes, but, technically, no. The reason that string pooling (deduping, caching, whatever you want to call it) is so successful in these benchmarks is because of the way that Joel constructs the test data. There is a 507KB file containing only 1695 rows, this file is duplicated until 1mil rows are provided. This means that there are only 1695 * 25 (#rows) = 42,375 unique strings that could possibly be needed. Actually even fewer, because there is a lot of duplication even within those 1695 rows. However, you might also observe that there are only 1695 unique *rows*; so could we get away with only allocating 1695 rows, and just repeating those in the List? I think that goes too far and takes too much advantage of fore-knowledge of the dataset. Honestly, I think Joel should forbid string caching too, as it also a bit abusive.

So, with 257MB as a baseline, it would seem that the fastest libraries are either producing (261-257) 4MB or (345 - 257) 88MB of additional allocations. I would estimate that a maximum of 1MB is needed for the string allocations, since with pooling enabled the maximum size would be 507KB (the size of the file) * 2 (decoding to UTF-16 strings). So, what accounts for the rest of the allocations? For the 261 class, the overhead is coming from List<T>. We don't specify a capacity when allocating the List. This means that the list will have to repeatedly grow to accommodate the 1mil items. Altering the above benchmark to not specify an initial capacity bumps the allocation to 260MB. So, the difference is about 1MB, which is about what I would expect.

So, what about the 345MB class? Both Cursively and RecordParser are hitting the same issue: there is a `Func<int,string>` that is being allocated for every row in the data set.

Here is where it's happening in RecordParser:

```
public List<T> GetRecords<T>(MemoryStream stream) where T : ICsvReadable, new()
        {
            var activate = ActivatorFactory.Create<T>(_activationMethod);
            var reader = BuildReader(activate);
            var result = ProcessStream<T>(stream, spanLine =>
            {
                var fields = reader.Parse(spanLine);
                var record = activate();

                record.Read(i => fields[i]); // <-- this is allocating a closure over fields

                return record;
            });

            return result;
        }
```

The inner lambda is allocating a closure over fields for each call to Read (1mil). You can refactor this to avoid that inner closure:

```
        public List<T> GetRecords<T>(MemoryStream stream) where T : ICsvReadable, new()
        {
            var activate = ActivatorFactory.Create<T>(_activationMethod);
            var reader = BuildReader(activate);

            string[] fields = Array.Empty<string>(); // <-- move fields outside of the func.
            Func<int, string> f = i => fields[i]; // <-- closure over fields only allocated once.

            var result = ProcessStream(stream, spanLine =>
            {
                fields = reader.Parse(spanLine); // <-- one more closure allocated here to capture fields.
                var record = activate();

                record.Read(f); // < f accesses fields via its own closure over the same variable

                return record;
            });

            return result;
        }
```

That takes RecordParser allocations down to 261, which is pretty much the minimum that can be expected.

Reply:
@MarkPflug thank you for your analysis, I found it very detailed! I didnt realize that the closure was being created for every loop/scope execution, great catch! I also could never imagine a closure in a loop could allocate 84MB. 

Thank you very much for all help!

Reply:
Just for curiosity, I try to allocated less than 261 and got it with the following change
Basically I use the first 1.000.000 records to calculate the average of record size, then combine it with the length of `MemoryStream` to have an expectation of existing records. So I can create the `List<>` setting the default capacity and avoid its grown. Allocations fall from 261 to 253 MB.

```c#
...
var reader = PipeReader.Create(stream);
var temp = new List<(long size, T item)>();

List<T> records = null;

while (true)
{
    ReadResult read = reader.ReadAsync().Result;
    ReadOnlySequence<byte> buffer = read.Buffer;
    while (TryReadLine(ref buffer, out ReadOnlySequence<byte> sequence))
    {
        var item = ProcessSequence(sequence, parser);

        if (records != null)
        {
            records.Add(item);
            continue;
        }

        temp.Add((sequence.Length, item));

        if (temp.Count == 1_000)
        {
            var avgSize = (int) temp.Average(x => x.size);
            var estimateRecordCount = stream.Length / avgSize;
            records = new List<T>((int)estimateRecordCount);

            foreach (var t in temp)
                records.Add(t.item);

            temp = null;
        }
    }
...
```